README.md
# NAME

Sque - Background job processing based on Resque, using Stomp

# VERSION

version 0.008

# SYNOPSIS

First you create a Sque instance where you configure the [Stomp](http://search.cpan.org/perldoc?Stomp)
backend and then you can start sending jobs to be done by workers:

    use Sque;

    my $s = Sque->new( stomp => '127.0.0.1:61613' );
    # Or, for failover
    $s = Sque->new( stomp => [ '127.0.0.1:61613', '127.0.0.2:61613' ] );

    $s->push( my_queue => {
        class => 'My::Task',
        args => [ 'Hello world!' ]
    });

You can also send by just using:

    $s->push({
        class => 'My::Task',
        args => [ 'Hello world!' ]
    });

In this case, the queue will be set automatically automatically to the
job class name with colons removed, which in this
case would be 'MyTask'.

Additionally, the [sque](http://search.cpan.org/perldoc?sque) command-line tool can be used to send messages:

    $ sque send -h 127.0.0.1 -p 61613 -c My::Task 'Hello world!'

Background jobs can be any perl module that implement a perform() function.
The [Sque::Job](http://search.cpan.org/perldoc?Sque::Job) object is passed as the only argument to this function:

    package My::Task;
    use strict;
    use 5.10.0;

    sub perform {
        my ( $job ) = @_;
        say $job->args->[0];
    }

    1;

Background jobs can also be OO.  The perform function will still be called
with the [Sque::Job](http://search.cpan.org/perldoc?Sque::Job) object as the only argument:

    package My::Task;
    use strict;
    use 5.10.0;
    use Moose;

    with 'Role::Awesome';

    has attr => ( is => 'ro', default => 'Where am I?' );

    sub perform {
        my ( $self, $job ) = @_;
        say $self->attr;
        say $job->args->[0];
    }

    1;

Finally, you run your jobs by instancing a [Sque::Worker](http://search.cpan.org/perldoc?Sque::Worker) and telling it
to listen to one or more queues:

    use Sque;

    my $w = Sque->new( stomp => '127.0.0.1:61613' )->worker;
    $w->add_queues('my_queue');
    $w->work;

Or you can simply use the [sque](http://search.cpan.org/perldoc?sque) command-line tool which uses [App::Sque](http://search.cpan.org/perldoc?App::Sque)
like so:

    $ sque work --host 127.0.0.1 --port 61613 --workers 5 --lib ./lib --lib ./lib2 --queues Queue1,Queue2,Queue3

# DESCRIPTION

This is a copy of [resque-perl](https://github.com/diegok/resque-perl)
by [Diego Kuperman](https://github.com/diegok) simplified a little bit
(for better or worse) and made to work with any stomp server rather than Redis.

# ATTRIBUTES

## stomp

A Stomp Client on this sque instance.

## namespace

Namespace for queues, default is 'sque'

## worker

A [Sque::Worker](http://search.cpan.org/perldoc?Sque::Worker) on this sque instance.

# METHODS

## push

Pushes a job onto a queue. Queue name should be a string and the
item should be a [Sque::Job](http://search.cpan.org/perldoc?Sque::Job) object or a hashref containing:
class - The String name of the job class to run.
args - Any arrayref of arguments to pass the job.

Example:

    $sque->push( archive => { class => 'Archive', args => [ 35, 'tar' ] } )

## pop

Pops a job off a queue. Queue name should be a string.
Returns a l<Sque::Job> object.

## key

Concatenate `$self-`namespace> with the received array of names
to build a redis key name for this sque instance.

## new_job

Build a [Sque::Job](http://search.cpan.org/perldoc?Sque::Job) object on this system for the given
hashref(see [Sque::Job](http://search.cpan.org/perldoc?Sque::Job)) or string(payload for object).

# ATTRIBUTES

# HELPER METHODS

# TODO

- Make App::Sque that will let you run sque and just pass it the
stomp server/port, queue list, lib directories (if needed), and
number of workers.
- More (real) tests.

# AUTHOR

William Wolf <throughnothing@gmail.com>

# COPYRIGHT AND LICENSE



William Wolf has dedicated the work to the Commons by waiving all of his
or her rights to the work worldwide under copyright law and all related or
neighboring legal rights he or she had in the work, to the extent allowable by
law.

Works under CC0 do not require attribution. When citing the work, you should
not imply endorsement by the author.

CHANGES
==================================================
Changes from 2011-11-15 00:00:00 +0000 to present.
==================================================

------------------------------------------
version 0.009 at 2012-11-14 22:04:04 +0000
------------------------------------------

  Change: 8d8917cb35bc0d5eaad569daad55b627dad34b66
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-11-14 13:29:38 +0000

    Add Travis.yml 

------------------------------------------
version 0.008 at 2012-11-14 19:36:20 +0000
------------------------------------------

  Change: c7232b8ec37b9efb14820e4e7ab000b0d4330027
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-11-14 11:36:20 +0000

    v0.008 

  Change: fe3d0eac2309038ff62e88355c9541d2c35e06f2
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-11-14 11:35:53 +0000

    Add support for stomp failover hosts 

  Change: dc130a8542ff360012cf2117d020a54541490fa8
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-10-29 12:23:28 +0000

    add TODO 

------------------------------------------
version 0.007 at 2012-07-16 17:11:43 +0000
------------------------------------------

  Change: 7deaffb44eefaa2b0f9e6903d603b68cb67d934a
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-07-16 10:11:43 +0000

    v0.007 

------------------------------------------
version 0.006 at 2012-05-02 16:25:27 +0000
------------------------------------------

  Change: d0e4ab6ceb4c8026b63d00211100fe5fb0432d41
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-05-02 12:25:27 +0000

    Add CHANGES 

  Change: 850ac2edeb936fa1211235d35b26d06114d76738
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-05-02 12:23:00 +0000

    Add more sleep to SpawnMQ

    This will give the tests more time to leat the mq.pl startup and get
    listening. I think that was causing test failures on a lot of systems
    where sleep 1 wasn't long enough. 

------------------------------------------
version 0.005 at 2012-04-25 04:06:35 +0000
------------------------------------------

  Change: 49457c7114605af6799322c79d2ca0ba33f735b6
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-04-24 21:06:35 +0000

    Merge pull request #3 from throughnothing/net-stomp

    Change Net::STOMP::Client => Net::Stomp 

  Change: ea42590e66545a3b645122230b407f9b260ba07c
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-04-24 23:59:54 +0000

    Add some more tests. 

  Change: dbfd381de286bab00a0ccf67557b866d992c1b8a
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-04-24 21:36:32 +0000

    Move to Net::Stomp from Net::STOMP::Client. 

  Change: 1e7d921047e4c4302657c310ccdabffea61bb23f
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-28 08:49:43 +0000

    Fix error in documentation 

------------------------------------------
version 0.004 at 2012-03-27 21:34:33 +0000
------------------------------------------

  Change: 6e89137e055a60ca680db10490ec1902e0d71fd7
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-27 14:34:33 +0000

    Merge pull request #2 from throughnothing/app

    Add sque command-line tool. 

  Change: 96d7c6dc484b68fe6dc3557f6b9e615f23835aad
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-27 17:32:58 +0000

    Add sque command-line tool.

    This adds the sque command-line tool with support for the work and
    send commands. It allows running a specified number of workers
    listening to a number of queues. The send command allows easy sending
    of messages to squeue workers. 

  Change: 9b1d9a81676bf9e31f3df55fe7e351c9c245f059
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-24 02:14:38 +0000

    Update docs 

  Change: f916d4fd64f432dd22c25a67833c99d72257103e
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-24 02:09:33 +0000

    Remove CHANGES file from repo, we don't need it 

  Change: ca8273b728f995ad742941a23cad850c65a23915
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-24 01:33:13 +0000

    Add TODO 

------------------------------------------
version 0.003 at 2012-03-18 02:25:38 +0000
------------------------------------------

  Change: 317a1efc33d4726606bf4ef8edd67abace2e5cd6
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-17 19:25:38 +0000

    v0.003 

  Change: 76753a2befd83952e1d0918575364e4645ca8940
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-17 19:23:48 +0000

    Merge pull request #1 from throughnothing/allow-oo-jobs

    Allow oo jobs 

  Change: 40d7f3e6180832defe656b4107054c8d6471687c
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-17 18:59:58 +0000

    Allow pushing to default queue.

    This allows you to not need to specify a queue on push and have it
    default to the job Class name with :'s replaced by -'s. 

  Change: 4a5bb312d8c23deace813030a41f709da7643866
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-17 04:33:04 +0000

    Allow OO jobs to be processed as well.

    This will allow much greater flexibility with jobs as you can have
    your jobs be full Moose classes that can share roles, etc. 

  Change: 2d5708dc3eaabbe7c0d4948b0109193caffe0e32
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-09 11:03:16 +0000

    doc typos 

------------------------------------------
version 0.002 at 2012-03-09 11:13:42 +0000
------------------------------------------

  Change: 951d9717396e1ae235cd64c67358cc83d6e6c539
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-09 03:13:42 +0000

    v0.002

    [William Wolf <throughnothing@gmail.com>]

    - update docs 

  Change: 5fcf52f82e07092ba016ffbd80d4ecde9483befd
  Author: William Wolf <throughnothing@gmail.com>
  Date : 2012-03-09 03:12:29 +0000

    update docs 

------------------------------------------
version 0.001 at 2012-03-09 10:37:19 +0000

lib/Sque.pm
use strict;
use warnings;
package Sque;
use Any::Moose;
use Any::Moose '::Util::TypeConstraints';
use Net::Stomp;

use Sque::Job;
use Sque::Worker;

# ABSTRACT: Background job processing based on Resque, using Stomp

subtype 'Sugar::Stomp' => as class_type('Net::Stomp');

coerce 'Sugar::Stomp'
    => from 'Str'
    => via {
        my ( $host, $port ) = split /:/;
        my $stomp = Net::Stomp->new({ hostname => $host, port => $port });
        $stomp->connect;
        return $stomp;
    };

coerce 'Sugar::Stomp'
    => from 'ArrayRef[Str]'
    => via {
        my ($a) = @_;
        my $hosts = [];
        for ( @$a ) {
            my ( $host, $port ) = split /:/;
            push @$hosts, { hostname => $host, port => $port };
        }

        my $stomp = @$hosts > 1
            ? Net::Stomp->new({ hosts => $hosts })
            : Net::Stomp->new({
                    hostname => $hosts->[0]{hostname},
                    port     => $hosts->[0]{port}
                });

        $stomp->connect;
        return $stomp;
    };

has stomp => (
    is => 'ro',
    lazy => 1,
    coerce => 1,
    isa => 'Sugar::Stomp',
    default => sub { Net::Stomp->new->connect },
);

has namespace => ( is => 'rw', default => sub { 'sque' });

has worker => (
    is => 'ro',
    lazy => 1,
    default => sub { Sque::Worker->new( sque => $_[0] ) },
);

sub push {
    my ( $self, $queue, $job ) = @_;
    confess "Can't push an empty job." unless ( $job || ref $queue );
    if( ref $queue ){
        $job = $self->new_job($queue) unless ref $queue eq 'Sque::Job';
        $queue = $job->queue;
    } else {
        $job = $self->new_job($job) unless ref $job eq 'Sque::Job';
    }

    $self->stomp->send( {
        persistent => 'true',
        destination => $self->key( $queue ),
        body => $job->encode,
    } );
}

sub pop {
    my ( $self ) = @_;
    my $frame = $self->stomp->receive_frame;
    return unless $frame;

    $self->new_job({
        frame => $frame,
        queue => $frame->destination,
    });
}

sub new_job {
    my ( $self, $job ) = @_;

    if ( $job && ref $job && ref $job eq 'HASH' ) {
        return Sque::Job->new({ sque => $self, %$job });
    }
    elsif ( $job ) {
        return Sque::Job->new({ sque => $self, payload => $job });
    }
    confess "Can't build an empty Sque::Job object.";
}

sub key {
    my $self = shift;
    '/queue/' . $self->namespace . '/' . shift;
}

__PACKAGE__->meta->make_immutable();

1;

=head1 SYNOPSIS

First you create a Sque instance where you configure the L<Stomp>
backend and then you can start sending jobs to be done by workers:

    use Sque;

    my $s = Sque->new( stomp => '127.0.0.1:61613' );
    # Or, for failover
    $s = Sque->new( stomp => [ '127.0.0.1:61613', '127.0.0.2:61613' ] );

    $s->push( my_queue => {
        class => 'My::Task',
        args => [ 'Hello world!' ]
    });

You can also send by just using:

    $s->push({
        class => 'My::Task',
        args => [ 'Hello world!' ]
    });

In this case, the queue will be set automatically automatically to the
job class name with colons removed, which in this
case would be 'MyTask'.

Additionally, the L<sque> command-line tool can be used to send messages:

    $ sque send -h 127.0.0.1 -p 61613 -c My::Task 'Hello world!'

Background jobs can be any perl module that implement a perform() function.
The L<Sque::Job> object is passed as the only argument to this function:

    package My::Task;
    use strict;
    use 5.10.0;

    sub perform {
        my ( $job ) = @_;
        say $job->args->[0];
    }

    1;

Background jobs can also be OO.  The perform function will still be called
with the L<Sque::Job> object as the only argument:

    package My::Task;
    use strict;
    use 5.10.0;
    use Moose;

    with 'Role::Awesome';

    has attr => ( is => 'ro', default => 'Where am I?' );

    sub perform {
        my ( $self, $job ) = @_;
        say $self->attr;
        say $job->args->[0];
    }

    1;

Finally, you run your jobs by instancing a L<Sque::Worker> and telling it
to listen to one or more queues:

    use Sque;

    my $w = Sque->new( stomp => '127.0.0.1:61613' )->worker;
    $w->add_queues('my_queue');
    $w->work;

Or you can simply use the L<sque> command-line tool which uses L<App::Sque>
like so:

    $ sque work --host 127.0.0.1 --port 61613 --workers 5 --lib ./lib --lib ./lib2 --queues Queue1,Queue2,Queue3

=head1 DESCRIPTION

This is a copy of L<resque-perl|https://github.com/diegok/resque-perl>
by L<Diego Kuperman|https://github.com/diegok> simplified a little bit
(for better or worse) and made to work with any stomp server rather than Redis.

=head1 ATTRIBUTES

=attr stomp

A Stomp Client on this sque instance.

=attr namespace

Namespace for queues, default is 'sque'

=attr worker

A L<Sque::Worker> on this sque instance.

=method push

Pushes a job onto a queue. Queue name should be a string and the
item should be a L<Sque::Job> object or a hashref containing:
class - The String name of the job class to run.
args - Any arrayref of arguments to pass the job.

Example:

    $sque->push( archive => { class => 'Archive', args => [ 35, 'tar' ] } )

=method pop

Pops a job off a queue. Queue name should be a string.
Returns a l<Sque::Job> object.

=head1 HELPER METHODS

=method key

Concatenate C<$self->namespace> with the received array of names
to build a redis key name for this sque instance.

=method new_job

Build a L<Sque::Job> object on this system for the given
hashref(see L<Sque::Job>) or string(payload for object).

=head1 TODO

=over 4

=item * Make App::Sque that will let you run sque and just pass it the
stomp server/port, queue list, lib directories (if needed), and
number of workers.

=item * More (real) tests.

=back

=cut
